tests/ssl-opt.sh

```diff
@@ -8007,6 +8007,24 @@ run_test    "ECJPAKE: working, TLS" \
             -S "SSL - The handshake negotiation failed" \
             -S "SSL - Verification of the message MAC failed"
 
+requires_config_enabled MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
+run_test    "ECJPAKE: working, TLS, opaque password" \
+            "$P_SRV debug_level=3 ecjpake_pw=bla ecjpake_pw_opaque=1" \
+            "$P_CLI debug_level=3 ecjpake_pw=bla ecjpake_pw_opaque=1\
+             force_ciphersuite=TLS-ECJPAKE-WITH-AES-128-CCM-8" \
+            0 \
+            -c "add ciphersuite: c0ff" \
```

Comment:
Please add a log that's exclusive to the new opaque version of the code path, as this one did not exercise it.

Reply:
I have a question on this point.
I implemented what you asked by adding a message to the `main()` functions of `ssl_server2/ssl_client2` and grepping it in `ssl-opt.sh` at runtime. At first it seemed to work perfectly, but then I faced issues with the CI, in particular with the `test_tls1_2_ecjpake_compatibility` test.

The problem in this test is that there are 2 different implementations involved: one with and the other without USE_PSA. As a consequence only one of the two is printing the "new" message, therefore causing `test_tls1_2_ecjpake_compatibility` to fail.
At the same time I don't think that adding a `requires_config_enabled` line to the `ssl-opt.sh` (for example with a check for `MBEDTLS_USE_PSA_CRYPTO`) would fix this since it checks the current configuration file, but we cannot tell if this was used to build `ssl_server2` or `ssl_client2`. Unless I change also the way `test_tls1_2_ecjpake_compatibility` executes the test, of course.

Do you have any suggestion on this?

Reply:
I would suggest having three tests here: one that checks these generic logs that ECJPAKE works; This will be the test that the `all.sh` compatibility component will use and run (you'll have to change the filter flags to only run this one).
Second test will use the non-PSA version (if ECJPAKE is enabled, regardless of USE_PSA);
Third test will use the opaque PSA version (depends_on ECJPAKE and USE_PSA).
This way all of the possibilities will be covered.
What do you think?

Reply:
I just pushed a [commit](https://github.com/Mbed-TLS/mbedtls/pull/6680/commits/a6b69dabc5edc13b3c7bb42233237652964c8ecd) which should partially address your request. In particular:
* it is possible to run compatibility test using PSA and non-PSA client/server (this is the old test)
* there's a new test specifically designed for the opaque password case

What's missing is the opportunity to launch either server and client with an invalid opaque password.
I'm working on this, but I think this will require to add another test in `all.sh`. The reason for this is that, with an invalid password, the server/client do not reach the "handshake" stage, therefore making `ssl-opt.sh` unusable. As a consequence the test should be done in `all.sh` in my opinion

Reply:
> For this purpose, wouldn't it be better to add a new test in test_suite_ssl.function instead?

Yes, I was precisely going to suggest that. Let's not forget that `ssl_{client,server}2` are just tools to test the library: our purpose is not to test the programs themselves. Sometimes the programs are the most convenient way to test the library (mostly when it's about a whole connection, especially interop with other programs), sometimes they're not.

Making sure that a single function rejects invalid inputs looks more like unit testing than integration testing, so is more likely to belong in `test_suite_ssl` than in `ssl-opt.sh`.

(Historical note: initially those programs were the only way we had of testing a full handshake. A few years ago `test_suite_ssl` has gained this ability as well. In general, if something can be conveniently tested either via the program or in `test_suite_ssl`, then `test_suite_ssl` is preferred as the tests execute much faster: no need to spawn processes, open sockets, etc, everything is in a single process and communications use a local buffer. However it's still often much more convenient to use the programs, so we still use them a lot.)

Reply:
Ah, I hadn't seen this comment when I wrote [that one](https://github.com/Mbed-TLS/mbedtls/pull/6680#issuecomment-1333408459) but I think we came to the same conclusion: option 2 is what we want.

Reply:
Generally speaking, when you find something you would test in a new function, and it's not tested in an existing function, chances are this was an oversight and we should add tests for both the old and the new function. (There are probably exceptions, but I think that's still a good rule of thumb.)

Reply:
Ok, I think I'd also like to see a test that uses an opaque password on server but non-opaque on client, the other way: this ensures that both methods handle the password in the same way. Just to be on the safe side, I'd like to see a case of password mismatch with opaque passwords.

Reply:
> The problem in this test is that there are 2 different implementations involved: one with and the other without USE_PSA.

Right, this is something that `ssl-opt.sh` can't understand right now. Ideally we would have things like `requires_config_enabled_server` and `requires_config_enabled_client` but that would be scope creep for this PR. I think what you did, using a specific name and `-e ECJPAKE_OPAQUE_PW` was probably the best compromise.